### ISSUE TITLE
React 18 concurrent instantiation and disposal

### ISSUE THREAD

**Issue by christianalfoni** [reactjs/react.dev#6283]
Hi there!

I have been reflecting on what React 18 concurrent mode means for instantiation and disposal of objects.

So something you could do before React 18 and StrictMode was:

```tsx
const SomeComponent = () => {
  const ref = useRef()

  if (!ref.current) {
    ref.current = new Foo()
  }

  useEffect(() => {
    return () => ref.current.dispose()
  }, [])
}
```

But as I understand, with React 18 (spearheaded by StrictMode) you can not really rely on creating instances during "rendering" cause React might call the function body multiple times and even abandon it before running any effects.

I understand that pure components are necessary for concurrent mode and to me it makes a lot of sense that React is becoming this pure state synchronisation using subscriptions type of implementation. But what does not make sense to me is that there is no way to "create something" when React has the intention of mounting it and "dispose that something" when React is abandoning the component due to concurrent mode or a plain unmount.

I wish there was some explicit documentation/statement that explains you can not rely on React 18 and component mounting/unmounting to control the instantiation and disposal of objects and more importantly some guides to how you should actually deal with it.

Which brings me to... have I gotten it wrong? Is there a way to safely do this? And if not, do we now rely on external state stores to determine when to instantiate/dispose of objects?

**Comment by christianalfoni** [comment]
So I found the following discussion: https://github.com/reactwg/react-18/discussions/18

The discussion gives a lot of insight into `useRef` and also use of a `getter` prop. But it is still not obvious to me how you would handle something like this.

As I understand React 18 can call the function bodies of components without applying anything to the DOM (no ref callbacks running) or running effects (no useEffect running). Then decide to just discard that at a later point? So if you need to instantiate something related to rendering components there does not seem to be any way to dispose of that if React decides to discard the rendering? 🤔

**Comment by gaearon** [comment]
Think of rendering as a calculation. The purpose of rendering is calculating a tree. There is no lifecycle associated with that by design. The lifecycle is *only* for attaching/detaching already calculated trees.

Can you say more about why _calculating_ a tree requires a resource? What is that resource? Concrete examples would help.

**Comment by christianalfoni** [comment]
Yes, sorry, so for example at CodeSandbox we have a resource (class) called `PitcherClient`, which is part of an SDK. This resource is responsible for talking to our VM process. We create this resource based on the current sandbox/repo you are consuming in the editor, where the sandbox/repo has a `projectId`.

The `PitcherClient` resource is created by a context provider and is exposed to several components which extracts state from this resource to display in the UI. We use the `projectId` as a ref on this context provider to remount it whenever you change sandbox/repo. We need to dispose of the resource when the component unmounts to disconnect from the current VM.

So simplified it looks something like this:

```tsx
function PitcherClientProvider({ projectId, children }) {
  const ref = useRef()

  if (!ref.current) {
    ref.current = new PitcherClient(projectId)
  }

  useEffect(() => {
    return () => ref.current.dispose()
  }, [])

  return <PitcherClientContext.Provider value={ref.current}>{children}</PitcherClientContext>
}

function ProjectPage({ projectId }) {
  return (
    <PitcherClientProvider key={projectId} projectId={projectId}>
      <ProjectEditor projectId={projectId} />
    </PitcherClientProvider>
  )
}
```

We are not attaching anything to the DOM with our context provider, so we can not use the `ref` callback to determine any attaching/detaching to the DOM.

As you can see we quickly fail in `StrictMode` here, but it works fine in production... at least for now.

We can jump hoops and create global state to create and dispose of the resource outside of Reacts reconciliation and use that same state to determine when the context provider should be rendered. But the lifecycle of this context provider IS the lifecycle of the resource... having this strong lifecycle coupling between a resource and a component is something I think makes us write more readable and predictable code, and has felt completely natural in React up to this point.

But this is not really me trying to say that React is doing something wrong, it is just really difficult to infer from the documentation when this code can fail in production. Cause currently it does not fail, it only fails with `StrictMode`.

**Comment by pascalgn** [comment]
I have a similar problem, aborting running HTTP requests also poses a problem in StrictMode.

Suppose I have the following component, everything is fine:

```jsx
function StatusComp() {
  const [state, setState] = useState("Fetching...");

  useEffect(() => {
    window.fetch("https://httpbin.io/delay/1").then(
      (res) => res.text().then((text) => setState(text)),
      (e) => setState(`${e}`)
    );
  }, []);

  return <div>State: {state}</div>;
};
```

Now I want to make sure that I cancel HTTP requests when the component gets unmounted. That way, I can put many components in a dialog and when the user closes the dialog via "Cancel" I can be sure that all running HTTP requests will be cancelled.

Cancelling the requests also ensures the user won't get a "Export finished!" message or so, many seconds after the dialog has already been closed.

```jsx
function StatusComp() {
  const fetch = useFetch();

  const [state, setState] = useState("");

  useEffect(() => {
    setState("Fetching...");
    fetch("https://httpbin.io/delay/1").then(
      (res) => res.text().then((text) => setState(text)),
      (e) => setState(`${e}`)
    );
  }, [fetch]);

  return <div>State: {state}</div>;
};

function useFetch() {
  const ref = useRef();
  if (!ref.current) {
    const abortController = new AbortController();
    const fetch = (request) =>
      window.fetch(request, { signal: abortController.signal });
    ref.current = { abortController, fetch };
  }

  useEffect(() => {
    return () => {
      ref.current && ref.current.abortController.abort("hook cleanup");
    };
  }, []);

  return ref.current.fetch;
};
```

This is exactly the same issue as described in the original comment, but with aborting HTTP requests, I think it's a common issue that many users might face. Using libraries like TanStack Query for data fetching doesn't even help, as they provide a way to cancel requests, but leave the actual triggering of the cancel up to the user.

@gaearon Any input would be highly appreciated! 🙏
